### PR TITLE
ref(typing): Type the events-trace endpoint

### PR DESCRIPTION
**Pull request by wmak**
- This updates the trace endpoint so that its passing typing again

### CODE REVIEW

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 1]
```diff
@@ -371,10 +401,10 @@ def full_dict(
         event_id = self.event["id"]
         # We're in a loop!
```
Comment (k-fish) [inline src/sentry/api/endpoints/organization_events_trace.py:402]: 🤔  Why is this comment here?

Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:402]: This refers to a loop in the trace structure lol

Comment (k-fish) [reply src/sentry/api/endpoints/organization_events_trace.py:402]: Do we really need this here?

Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:400]: Hmm probably could reword it to be better

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 2]
```diff
@@ -41,124 +45,144 @@
...
-        "precise.start_ts": int,
+        "measurements": dict[str, int],
+        "occurrence_id": list[str],
+        "occurrence_spans": list[SnubaSpan],
+        "occurrence_to_issue_id": dict[str, list[int]],
         "precise.finish_ts": int,
-        "trace.span": str,
+        "precise.start_ts": int,
+        "profile.id": str,
+        "profile.profiler_id": str,
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:71]: The type here changed from before from

```python
"profile.profile_id" -> "profile.profiler_id"
"profiler.id" -> "profile.id"
```

Can you double check to make sure which one is correct?

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 3]
```diff
@@ -371,10 +399,10 @@ def full_dict(
         event_id = self.event["id"]
         # We're in a loop!
         if event_id in visited:
-            return
+            return None
         else:
             visited.add(self.event["id"])
-        result = self.to_dict()
+        result = cast(FullResponse, self.to_dict())
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:405]: Is this casting a `LightResponse` to a `FullResponse`? Is this right?

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 4]
```diff
@@ -1013,11 +1053,6 @@ def get(self, request: HttpRequest, organization: Organization, trace_id: str) -
         if event_id and not is_event_id(event_id):
             return Response({"detail": INVALID_ID_DETAILS.format("Event ID")}, status=400)
 
-        tracing_without_performance_enabled = features.has(
-            "organizations:performance-tracing-without-performance",
-            organization,
-            actor=request.user,
-        )
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:1020]: Is this flag fully GA'ed now? Are there environments where this can't be turned on like single tenant or self hosted?

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 5]
```diff
@@ -1063,11 +1096,23 @@ def get(self, request: HttpRequest, organization: Organization, trace_id: str) -
...
+        limit: int,
+        transactions: Sequence[SnubaTransaction],
+        errors: Sequence[SnubaError],
+        roots: Sequence[SnubaTransaction],
+        warning_extra: dict[str, str],
+        event_id: str | None,
+        detailed: bool = False,
+        use_spans: bool = False,
+    ) -> Any:
+        pass
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:1114]: Is this an abstract method that is overridden in the child classes? If so, can it

```python
raise NotImplementedError
```

Bonus points if you turn this into an abstract class and decorate this with `@abc.abstractmethod`

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 6]
```diff
@@ -1158,7 +1195,7 @@ def serialize(
 
         with sentry_sdk.start_span(op="building.trace", description="light trace"):
             # Check if the event is an orphan_error
-            if not snuba_event and not nodestore_event and allow_orphan_errors:
+            if not snuba_event or not nodestore_event:
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:1198]: Why did this `and` change into an `or`?

File: src/sentry/api/endpoints/organization_events_trace.py [hunk 7]
```diff
@@ -1325,9 +1358,11 @@ def serialize(
                 warning_extra,
                 event_id,
                 detailed,
-                allow_orphan_errors,
             )
             return results
+
+        # Code past here is deprecated, but must continue to exist until sentry installs in every possible environment
+        # are able to support storing span data
```
Comment (Zylphrex) [inline src/sentry/api/endpoints/organization_events_trace.py:1365]: We need to be more strict here. As long as any environment has the option to turn off spans, we have to keep this code around.

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 2 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:71]: New one is correct, these types weren't being used before so they were just wrong without being noticed

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 3 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:405]: Yeah, I a FullResponse shares a lot with a LightResponse so this is just extending it. Probably could rewrite this to be a bit cleaner, but we aren't using light responses anymore in environments where spans are enabled

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 4 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:1020]: Fully GA'ed, the frontend already assumes this flag is on since it changes the response 😬

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 5 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:1114]: fiiiine

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 6 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:1198]: 👀 uh. no idea good catch

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 7 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:1365]: Hmm what do you mean by being more strict?

Comment (Zylphrex) [reply src/sentry/api/endpoints/organization_events_trace.py:1365]: I guess what I'm trying to say is that it is not enough that every environment is capable of storing span data. Every environment must only use span data before this old code can be removed.

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 6 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py:1198]: Oh nevermind, this is a typing catch, the rest of this function needs both snuba_event and nodestore_event to be defined, so if either isn't defined we want to do the following block

File: src/sentry/api/endpoints/organization_events_trace.py (hunk 7 above)
Comment (wmak) [reply src/sentry/api/endpoints/organization_events_trace.py]: Reworded this slightly

**Review by k-fish — APPROVED** [review]
thanks for the cleanup, nice types Will 👏
